### PR TITLE
KML module to generate KML files from waypoints. 

### PR DESCRIPTION
**Pull request by wuda17**
Ticket link: https://app.asana.com/0/0/1205055426695845/f

### CODE REVIEW

File: kml/modules/waypoints_to_kml.py [hunk 1]
```diff
@@ -0,0 +1,46 @@
+"""
+Module used to convert list of waypoints to KML file
```
Comment (Xierumeng) [inline kml/modules/waypoints_to_kml.py:2]: Nit: Period.

File: kml/modules/waypoints_to_kml.py [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+"""
+import os
```
Comment (Xierumeng) [inline kml/modules/waypoints_to_kml.py:4]: Use `pathlib` .

File: kml/modules/waypoints_to_kml.py [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+
+import simplekml
+
+
+def waypoints_to_kml(waypoints: "list[tuple[float, float]]",
+                     document_name: str,
+                     save_path: str = None) -> str:
```
Comment (Xierumeng) [inline kml/modules/waypoints_to_kml.py:11]: Use `pathlib` paths.

Comment (Xierumeng) [inline kml/modules/waypoints_to_kml.py:11]: Return bool for success/failure.

Comment (Xierumeng) [inline kml/modules/waypoints_to_kml.py:11]: Don't use default argument.

File: kml/modules/waypoints_to_kml.py [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+
+    Returns
+    -------
+    str
+        Path to generated KML file.
+    """
+    kml = simplekml.Kml()
+
+    for idx, waypoint in enumerate(waypoints):
+        waypoint_name = f'Point {idx}'
```
Comment (Xierumeng) [inline kml/modules/waypoints_to_kml.py:33]: Use double quotation marks for strings.

File: kml/test_waypoints_to_kml.py [hunk 5]
```diff
@@ -0,0 +1,84 @@
+"""
+Test process for KML generation
+"""
+import os
```
Comment (Xierumeng) [inline kml/test_waypoints_to_kml.py:4]: Use `pathlib` .

File: kml/test_waypoints_to_kml.py [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+
+import pytest
+from pykml import parser
```
Comment (Xierumeng) [inline kml/test_waypoints_to_kml.py:7]: I am not comfortable with bringing in another similar library for testing.

File: kml/modules/waypoints_to_kml.py [hunk 7]
```diff
@@ -0,0 +1,42 @@
+import simplekml
```
Comment (Xierumeng) [inline kml/modules/waypoints_to_kml.py:1]: Add file docstring.

File: kml/modules/waypoints_to_kml.py [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+from pathlib import Path
```
Comment (Xierumeng) [inline kml/modules/waypoints_to_kml.py:2]: System level imports, followed by empty line, followed by package level imports.

Comment (Xierumeng) [inline kml/modules/waypoints_to_kml.py:2]: Import namespace rather than class: `import pathlib`

File: kml/modules/waypoints_to_kml.py [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+
+    Parameters
+    ----------
+    waypoints: list[tuple[float, float]]
+        Waypoint coordinates in decimal degrees (latitude, longitude).
+    document_name: str
+        Name of the KML file to save (without the .kml extension).
+    save_path: pathlib.Path, optional
+        Path where the KML file should be saved. If not provided, the file
+        will be saved in the current working directory.
```
Comment (Xierumeng) [inline kml/modules/waypoints_to_kml.py:19]: Update this, doesn't appear to be optional.

File: kml/modules/waypoints_to_kml.py [hunk 10]
```diff
@@ -0,0 +1,42 @@
...
+    kml = simplekml.Kml()
+
+    for idx, waypoint in enumerate(waypoints):
+        waypoint_name = f"Point {idx}"
+        lat, lng = waypoint
+
+        # coords parameters are in the order: lon, lat, optional height
+        kml.newpoint(name=waypoint_name, coords=[(lng, lat)])
+
+    kml_file_path = save_path / f'{document_name}.kml'
```
Comment (Xierumeng) [inline kml/modules/waypoints_to_kml.py:35]: Use `pathlib.Path()` constructor.

Comment (Xierumeng) [inline kml/modules/waypoints_to_kml.py:35]: Use double quotation marks for strings.

File: kml/test_waypoints_to_kml.py [hunk 11]
```diff
@@ -0,0 +1,43 @@
+import os
```
Comment (Xierumeng) [inline kml/test_waypoints_to_kml.py:1]: Add file docstring.

Comment (Xierumeng) [inline kml/test_waypoints_to_kml.py:1]: Remove unused import.

File: kml/test_waypoints_to_kml.py [hunk 12]
```diff
@@ -0,0 +1,43 @@
...
+from pathlib import Path
```
Comment (Xierumeng) [inline kml/test_waypoints_to_kml.py:2]: Import namespace.

File: kml/test_waypoints_to_kml.py [hunk 13]
```diff
@@ -0,0 +1,43 @@
...
+
+from modules import waypoints_to_kml
+
+
+@pytest.fixture
+def waypoints():
+    return [(37.7749, -122.4194), (34.0522, -118.2437), (40.7128, -74.0060)]
+
+
+def test_waypoints_to_kml_with_save_path(waypoints, tmp_path):
```
Comment (Xierumeng) [inline kml/test_waypoints_to_kml.py:14]: Type hinting.

File: kml/test_waypoints_to_kml.py [hunk 14]
```diff
@@ -0,0 +1,43 @@
...
+    """
+    Basic test case to save KML to the correct path when provided.
+    """
+    actual_document_name = "actual_kml_document"
+
+    # Build a temporary directory using tmp_path so the KML files are cleaned after the tests are run
+    save_path = tmp_path / "path" / "to" / "save"
```
Comment (Xierumeng) [inline kml/test_waypoints_to_kml.py:21]: Use `pathlib.Path()` constructor.

File: kml/test_waypoints_to_kml.py [hunk 15]
```diff
@@ -0,0 +1,43 @@
...
+    save_path.mkdir(parents=True, exist_ok=True)
+
+    kml_status_success = waypoints_to_kml.waypoints_to_kml(waypoints, actual_document_name, save_path)
```
Comment (Xierumeng) [inline kml/test_waypoints_to_kml.py:24]: Rename to `result`

File: kml/test_waypoints_to_kml.py [hunk 16]
```diff
@@ -0,0 +1,43 @@
...
+
+    # Assert that the operation was a success
+    assert kml_status_success is True
```
Comment (Xierumeng) [inline kml/test_waypoints_to_kml.py:27]: Remove `is True` (it is not necessary).

**Review by Xierumeng — COMMENTED** [review]
Reviewed.

File: kml/test_waypoints_to_kml.py [hunk 17]
```diff
@@ -0,0 +1,43 @@
+"""
+Test Process.
+"""
+
+import pathlib
+import pytest
```
Comment (Xierumeng) [inline kml/test_waypoints_to_kml.py:6]: System level imports, followed by empty line, followed by package level imports.

File: kml/test_waypoints_to_kml.py [hunk 18]
```diff
@@ -0,0 +1,43 @@
+"""
+Test Process.
+"""
+
```
Comment (Xierumeng) [inline kml/test_waypoints_to_kml.py:4]: Remove empty line.

File: kml/test_waypoints_to_kml.py [hunk 19]
```diff
@@ -0,0 +1,43 @@
...
+from modules import waypoints_to_kml
+
+
+@pytest.fixture
+def waypoints():
+    return [(37.7749, -122.4194), (34.0522, -118.2437), (40.7128, -74.0060)]
+
+
+def test_waypoints_to_kml_with_save_path(waypoints: "list[tuple[float, float]]",
+                                         tmp_path):
```
Comment (Xierumeng) [inline kml/test_waypoints_to_kml.py:17]: Type hint: `pathlib.Path` as specified in the Pytest documentation.

File: kml/modules/waypoints_to_kml.py [hunk 20]
```diff
@@ -0,0 +1,45 @@
...
+Module to convert waypoints list to kml document.
+"""
+import pathlib
+
+import simplekml
+
+
+def waypoints_to_kml(waypoints: "list[tuple[float, float]]",
+                     document_name: str,
+                     save_path: pathlib.Path) -> bool:
```
Comment (Xierumeng) [inline kml/modules/waypoints_to_kml.py:11]: Rename to `save_directory`

File: kml/test_waypoints_to_kml.py [hunk 21]
```diff
@@ -0,0 +1,43 @@
...
+    """
+    Basic test case to save KML to the correct path when provided.
+    """
+    actual_document_name = "actual_kml_document"
+
+    # Build a temporary directory using tmp_path so the KML files are cleaned after the tests are run
+    save_path = pathlib.Path(tmp_path, "path", "to", "save")
```
Comment (Xierumeng) [inline kml/test_waypoints_to_kml.py:24]: Rename to `save_directory`

Comment (Xierumeng) [inline kml/test_waypoints_to_kml.py:24]: Is path/to/save necessary? Why not just the original Pytest temporary directory?

File: kml/test_waypoints_to_kml.py [hunk 22]
```diff
@@ -0,0 +1,43 @@
...
+    save_path.mkdir(parents=True, exist_ok=True)
+
+    result = waypoints_to_kml.waypoints_to_kml(waypoints, actual_document_name, save_path)
```
Comment (Xierumeng) [inline kml/test_waypoints_to_kml.py:27]: Assert on result.

File: kml/test_waypoints_to_kml.py [hunk 23]
```diff
@@ -0,0 +1,43 @@
...
+
+    # Assert that the KML file has been generated properly in the provided path
+    assert kml_file_path.exists()
+    assert kml_file_path.suffix == ".kml"
+
+    # Get the directory where the test file is located
+    test_directory = pathlib.Path(__file__).parent
+
+    # Define the path to the static KML file for comparison (relative to the test file directory)
+    static_kml_path = test_directory / "expected_document.kml"
```
Comment (Xierumeng) [inline kml/test_waypoints_to_kml.py:40]: Use constructor.

**Review by Xierumeng — COMMENTED** [review]
Reviewed.

File: kml/test_waypoints_to_kml.py [hunk 24]
```diff
@@ -0,0 +1,45 @@
...
+    """
+    actual_document_name = "actual_kml_document"
+
+    # Build a temporary directory using tmp_path so the KML files are cleaned after the tests are run
+    tmp_path.mkdir(parents=True, exist_ok=True)
+
+    result = waypoints_to_kml.waypoints_to_kml(waypoints, actual_document_name, tmp_path)
+
+    # Assert success
+    assert result is True
```
Comment (Xierumeng) [inline kml/test_waypoints_to_kml.py:29]: Nit: `assert result`

File: kml/test_waypoints_to_kml.py [hunk 25]
```diff
@@ -0,0 +1,45 @@
...
+
+    # Assert that the KML file has been generated properly in the provided path
+    assert kml_file_path.exists()
+    assert kml_file_path.suffix == ".kml"
+
+    # Get the directory where the test file is located
+    test_directory = pathlib.Path(__file__).parent
+
+    # Define the path to the static KML file for comparison (relative to the test file directory)
+    static_kml_path = pathlib.Path(test_directory, "expected_document.kml")
```
Comment (Xierumeng) [inline kml/test_waypoints_to_kml.py:42]: Make the expected document name a global constant.

Comment (Xierumeng) [inline kml/test_waypoints_to_kml.py:42]: I'm not sure if this is necessary, does `pathlib.Path("expected_document.kml")` work as well?

**Review by Xierumeng — COMMENTED** [review]
Reviewed.

File: kml/test_waypoints_to_kml.py [hunk 26]
```diff
@@ -0,0 +1,48 @@
...
+
+    # Assert that the KML file has been generated properly in the provided path
+    assert kml_file_path.exists()
+    assert kml_file_path.suffix == ".kml"
+
+    # Get the directory where the test file is located
+    test_directory = pathlib.Path(__file__).parent
+
+    # Define the path to the static KML file for comparison (relative to the test file directory)
+    static_kml_path = test_directory / EXPECTED_KML_DOCUMENT_NAME
```
Comment (Xierumeng) [inline kml/test_waypoints_to_kml.py:45]: What I meant was: Does `static_kml_path = pathlib.Path(EXPECTED_KML_DOCUMENT_NAME)` work? As in do you actually need to get the full parent directory?

**Review by Xierumeng — COMMENTED** [review]
Reviewed.

File: kml/test_waypoints_to_kml.py (hunk 26 above)
Comment (wuda17) [reply kml/test_waypoints_to_kml.py:45]: Oh I understand what you're asking now, yes, it does the changes that were made.

**Review by Xierumeng — APPROVED** [review]
Approved.

**Pull request merged**